### PR TITLE
Enable additional_code support & update templates

### PR DESCRIPTION
**Pull request by kaiming-cheng**
Context:
We are adding additional_code to improve the overall performance of Triton KernelAgent 

Changes:
- Add additional_code parameter to TritonKernelAgent.generate_kernel()
- Update agent.py to pass additional_code through test generation and kernel seed generation
- Update prompt_manager.py to pass additional_code to render_test_generation_prompt, render_test_generation_prompt, and render_kernel_refinement_prompt
- Update manager.py and worker.py to support additional_code in the verification and refinement loop
- Update templates (kernel_generation.j2, kernel_refinement.j2, test_generation.j2) based on the perf implementations
- Add example demonstrating additional_code usage with dropout_residual

Additional_code is optional and compatible with existing code.

Test:
`python triton_04_fused_dropout_residual.py --additional-code dropout_residual/additional_code.py `
✅ SUCCESS! Generated kernel in 1 rounds

## Benchmark triton_04 with additional_code Results

**Specs:** Batch size: 32, Sequence length: 2048, Hidden dimension: 4096

### Without Additional Code

| Run | Triton Kernel (ms) | PyTorch Native (ms) | Speedup |
|-----|-------------------|---------------------|---------|
| 1   | 2.8583            | 11.7985             | 4.13x   |
| 2   | 2.3589            | 3.3123              | 1.40x   |
| 3   | 4.0277            | 12.9672             | 3.22x   |
| 4   | 2.5105            | 3.3045              | 1.32x   |

### With Additional Code

| Run | Triton Kernel (ms) | PyTorch Native (ms) | Speedup |
|-----|-------------------|---------------------|---------|
| 1   | 2.3608            | 3.3090              | 1.40x   |
| 2   | 2.5113            | 3.3045              | 1.32x   |
| 3   | 2.4178            | 3.3083              | 1.37x   |
| 4   | 2.3584            | 3.3097              | 1.40x   |

### Summary Statistics

| Metric | Without Additional Code | With Additional Code |
|--------|------------------------|---------------------|
| Avg Triton Kernel | 2.9389 ms | 2.4121 ms |
| Avg PyTorch Native | 7.8456 ms | 3.3079 ms |

### CODE REVIEW

File: templates/kernel_refinement.j2 [hunk 1]
```diff
@@ -57,6 +80,11 @@ CRITICAL REQUIREMENTS - NO CHEATING ALLOWED:
 6. Focus on fixing the specific error while maintaining correctness
 7. Prefer solutions that preserve or extend operation fusion; never intentionally unfuse a pipeline unless you document why it is impossible to keep it fused
 8. Keep the wrapper free of PyTorch compute primitives—no torch.nn modules, torch.nn.functional calls, or other high-level PyTorch ops in the execution path
```
Comment (Jack-Khuu) [inline templates/kernel_refinement.j2:82]: Not an issue, just flagging delta (new instructions here, not in perf branch)

File: examples/triton_04_fused_dropout_residual.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+    print(f"Working directory: {base_dir}")
+    assert os.path.exists(base_dir), f"Directory {base_dir} does not exist"
+
+    path = os.path.join(base_dir, "additional_code.py")
+    if os.path.exists(path):
+        with open(path, "r") as f:
+            additional_code = f.read()
+
+    return additional_code
```
Comment (Jack-Khuu) [inline examples/triton_04_fused_dropout_residual.py]: Good code, but can we just try catch in main if we're just expecting one arg path

File: examples/triton_04_fused_dropout_residual.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    print("Fused dropout-residual Operation Kernel Generation")
+    print("Dropout rate: 0.1")
+    print(
+        "Input tensors shape: batch size: 32, sequence length: 2048, hidden dimension: 4096"
+    )
+    print("Operation: dropout + residual")
+    print("=" * 80)
+
+    # Define the problem
+    problem_description = """
```
Comment (Jack-Khuu) [inline examples/triton_04_fused_dropout_residual.py:73]: relation between this and the one in `drop_residual/`?

File: examples/dropout_residual/problem_description.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+Create a Triton kernel for fused dropout + residual addition.
+
+The kernel should:
+1. Apply dropout to input tensor with probability p
+2. Add the result to a residual tensor
+3. Return the final output
+
+Input tensors shape: (batch_size, seq_len, hidden_dim)
+All tensors are float32 on CUDA device.
```
Comment (Jack-Khuu) [inline examples/dropout_residual/problem_description.md]: Not necessarily a problem, just flagging the new line whitespace added here

File: templates/test_generation.j2 [hunk 5]
```diff
@@ -18,111 +18,146 @@ Generate a comprehensive test for a Triton kernel based on the following problem
 
 {{ problem_description }}
```
Comment (Jack-Khuu) [inline templates/test_generation.j2:19]: This file is significantly different in the branch/here than on main. Will defer to @Laurawly

File: templates/kernel_generation.j2 [hunk 6]
```diff
@@ -53,11 +76,12 @@ CRITICAL REQUIREMENTS - NO CHEATING ALLOWED:
 - The actual computation MUST be implemented using Triton operations (tl.load, tl.store, tl.sum, etc.)
 - The wrapper function can use PyTorch for tensor creation, memory allocation, and result formatting, but the core computation MUST happen in the Triton kernel
 - DO NOT import or instantiate torch.nn modules, call torch.nn.functional (including aliases like F.*), or use PyTorch activations/pooling helpers to satisfy the requirements
```
Comment (Jack-Khuu) [inline templates/kernel_generation.j2:78]: Not an issue, just flagging delta (new instructions here, not in perf branch)

File: templates/kernel_generation.j2 [hunk 7]
```diff
@@ -66,6 +90,10 @@ CRITICAL REQUIREMENTS - NO CHEATING ALLOWED:
 7. Handle edge cases and boundary conditions properly
 8. Explicitly reason about what is being fused inside the wrapper docstring or leading comments so reviewers can confirm you attempted fusion
 9. Keep the wrapper free of PyTorch compute primitives—no torch.nn modules, torch.nn.functional calls, or other high-level PyTorch operators in the execution path
```
Comment (Jack-Khuu) [inline templates/kernel_generation.j2:92]: Not an issue, just flagging delta (new instructions here, not in perf branch)

File: examples/triton_04_fused_dropout_residual.py (hunk 3 above)
Comment (kaiming-cheng) [reply examples/triton_04_fused_dropout_residual.py:73]: Previous example.py reads problem description from the `drop_residual`. I moved the description to the main for consistency with other examples. We can remove the .md file in the `drop_residual` folder

File: templates/kernel_generation.j2 (hunk 7 above)
Comment (kaiming-cheng) [reply templates/kernel_generation.j2:92]: > Not an issue, just flagging delta (new instructions here, not in perf branch)

These are the instructions included in the main branch. The instructions from the perf branch are added on top of the main branch if not explicitly removed.

File: templates/kernel_refinement.j2 (hunk 1 above)
Comment (kaiming-cheng) [reply templates/kernel_refinement.j2:82]: > Not an issue, just flagging delta (new instructions here, not in perf branch)

These are the instructions included in the main branch. The instructions from the perf branch are added on top of the main branch if not explicitly removed.

File: templates/test_generation.j2 (hunk 5 above)
Comment (kaiming-cheng) [reply templates/test_generation.j2:19]: > This file is significantly different in the branch/here than on main. Will defer to @Laurawly

Instructions here in main are explicitly removed in the perf branch

File: examples/triton_04_fused_dropout_residual.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+from dotenv import load_dotenv
+from triton_kernel_agent import TritonKernelAgent
+
+
+def main():
+    """Generate and test a fused dropout-residual kernel."""
+
+    # Load additional code if working directory provided
+    parser = argparse.ArgumentParser(
```
Comment (Jack-Khuu) [inline examples/triton_04_fused_dropout_residual.py:33]: Ah I should've clarified, since this is just an example let's just expect a direct path to the `additional_code.py` instead of a working dir
- Or conversely is there a reason why `additional_code` can't also be inlined?


``` python
if args.additional_code:
  try:
     ...
  catch:
     ...
else:
     ...
```
